fgpyo/sam/__init__.py

```diff
@@ -1036,55 +1040,96 @@ class ReadEditInfo:
         reference_offset: if provided, assume that reference_sequence[reference_offset] is the
             first base aligned to in reference_sequence, otherwise use r.reference_start
 
     Returns:
         a ReadEditInfo with information about how the read differs from the reference
     """
-    assert not rec.is_unmapped, f"Cannot calculate edit info for unmapped read: {rec}"
+    if rec.is_unmapped or rec.query_sequence is None or rec.query_sequence == NO_QUERY_BASES:
+        return None
 
-    query_offset = 0
-    target_offset = reference_offset if reference_offset is not None else rec.reference_start
-    cigar = Cigar.from_cigartuples(rec.cigartuples)
+    query_offset: int = 0
+    target_offset: int = reference_offset if reference_offset is not None else rec.reference_start
+    cigar: Cigar = Cigar.from_cigartuples(rec.cigartuples)
 
     matches, mms, insertions, ins_bases, deletions, del_bases = 0, 0, 0, 0, 0, 0
-    ok_bases = {"A", "C", "G", "T"}
-
+    ok_bases: set[str] = {"A", "C", "G", "T"}
+    md_edits: list[str] = []
+    current_match_count: int = 0
     for elem in cigar.elements:
         op = elem.operator
-
-        if op == CigarOp.I:
-            insertions += 1
-            ins_bases += elem.length
-        elif op == CigarOp.D:
+        # TODO: use match-case statements after we drop Python 3.9 support
+        if op == CigarOp.N:  # skipped region from ref, consumes ref
+            target_offset += elem.length
+        elif op in (CigarOp.I, CigarOp.S):  # consumes query
+            query_offset += elem.length
+            if op == CigarOp.I:
+                insertions += 1
+                ins_bases += elem.length
+        elif op == CigarOp.D:  # consumes ref
+            if target_offset + elem.length >= len(reference_sequence):
+                break  # out of bounds
             deletions += 1
             del_bases += elem.length
-        elif op == CigarOp.M or op == CigarOp.X or op == CigarOp.EQ:
-            for i in range(0, elem.length):
-                q = rec.query_sequence[query_offset + i].upper()
-                t = reference_sequence[target_offset + i].upper()
-                if q != t or q not in ok_bases:
+            md_edits.append(str(current_match_count))  # append match count and reset
+            current_match_count = 0
+            md_edits.append(f"^{reference_sequence[target_offset:target_offset + elem.length]}")
+            target_offset += elem.length
+        elif op in (CigarOp.M, CigarOp.X, CigarOp.EQ):
+            for in_block_offset in range(0, elem.length):
+                if (target_offset + in_block_offset) >= len(reference_sequence):
+                    break  # out of bounds
+                query_base: str = rec.query_sequence[query_offset + in_block_offset].upper()
+                ref_base: str = reference_sequence[target_offset + in_block_offset].upper()
+
+                if query_base != ref_base or (query_base not in ok_bases and not n_as_match):
```

Comment:
issue:

To support that `=` is always a match, regardless of the reference.

```suggestion
                if (query_base != ref_base and query_base != "=") or (query_base not in ok_bases and not n_as_match):
```

Reply:
fixed